### PR TITLE
Add memoization per context

### PR DESCRIPTION
**Pull request by osdnk**

### CODE REVIEW

**Comment by osdnk** [comment]
I'll follow up with java

**Comment by osdnk** [comment]
Fixed style (4 spaces -> 2)

Removed useless "params" in function node

Removed name from params (not used)

Removed unused imports

File: ios/Nodes/REANode.m [hunk 1]
```diff
@@ -65,11 +67,16 @@ - (id)evaluate
...
-    _lastLoopID = _updateContext.loopID;
-    return (_memoizedValue = [self evaluate]);
+  if (![_lastLoopID objectForKey:_updateContext.callID] || [[_lastLoopID objectForKey:_updateContext.callID] longValue] < [_updateContext.loopID longValue]) {
+    [_lastLoopID setObject:_updateContext.loopID forKey:_updateContext.callID];
+    id val = [self evaluate];
+    if (val == 0) {
+      val = [[NSNumber alloc] initWithInt:0];
+    }
+    [_memoizedValue setObject:val forKey:self.updateContext.callID];
+    return val;
```
Comment (chrfalch) [inline ios/Nodes/REANode.m:77]: I'm having some issue with getting the wrong values back from the function nodes. It might be that things happen in the "wrong" order here:

1) BeginContext, updateContext.callID is updated
2) Function Call evaluates AND pops the CallID
3) _memoizedValue is set for the wrong callID on line 76...

This is not evident in the example in Reanimated, but I'm seeing it "loud and clear" in my animation examples.

Comment (chrfalch) [reply ios/Nodes/REANode.m:77]: In my example I'm calling the same proc multiple times each loop, and I can see that I'm getting the same value from calling the same function twice with different parameters.

Comment (osdnk) [reply ios/Nodes/REANode.m:77]: @chrfalch 
`_memoizedValue` is set inside the evaluation, no? Could you add some breaking example?

Comment (osdnk) [reply ios/Nodes/REANode.m:77]: I probably see the problem. Maybe the reason is that we should drop revert the context while evaluating param? Actually inout of param ought to be evaluated in previous context, no?

File: ios/Nodes/REAParamNode.m [hunk 2]
```diff
@@ -3,37 +3,35 @@
...
+- (void)endContext {
+  [_argstack removeLastObject];
 }
 
 
--(id) evaluate {
-    REANode * node = [self.nodesManager findNodeByID:[_argstack lastObject]];
-    return [node value];
+- (id)evaluate {
+  REANode * node = [self.nodesManager findNodeByID:[_argstack lastObject]];
```
Comment (osdnk) [inline ios/Nodes/REAParamNode.m:44]: probably here we need to go back to the previous context (set `prevCallId` as callID) and after evaluation  revert this change.

File: ios/Nodes/REANode.m (hunk 1 above)
Comment (chrfalch) [reply ios/Nodes/REANode.m:77]: Yes, I think that seems to be the issue. The result of evaluating multiple functions in the same iteration seems to return the same value.

Comment (chrfalch) [reply ios/Nodes/REANode.m:77]: I have tested and it seems to exhibit the same problems as before unfortunately.

**Comment by chrfalch** [comment]
Ok, so I think I know what's happening here. The problem is with updating nodes using the `set` node inside a function - the memoized value will be set with the proc node's callID which will be wrong when trying to evaluate the node from outside the function.

This is a simple reproduction (paste into test.js):

```js
import React, { Component } from 'react';
import { View } from 'react-native';
import Animated from 'react-native-reanimated';

const { set, call, block, proc } = Animated;

const updateValue = proc((a) => block([
  set(a, 1000),  
]));

export default class Example extends Component {
  constructor(props) {
    super(props);
    this.value1 = new Animated.Value(100);    
    this.value2 = new Animated.Value(100);    
  }

  render() {
    return (
      <View>
        <Animated.Code>
          {() => block([
            // First version without proc
            call(
              [this.value1],
              ([value]) => {
                console.log("this.value before update:", value);
              }
            ),
            set(this.value1, 1000),
            call(
              [this.value1],
              ([value]) => {
                console.log("this.value after update:", value);
              }
            ),
            // With proc
            call(
              [this.value2],
              ([value]) => {
                console.log("this.value before proc:", value);
              }
            ),
            updateValue(this.value2),
            call(
              [this.value2],
              ([value]) => {
                console.log("this.value after proc:", value);
              }
            )])          
          }
        </Animated.Code>       
      </View>
    );
  }
}
```

The output will be:

```
this.value before update: 100
this.value after update: 1000
this.value before proc: 100
this.value after proc: 100
```

Expected output:
```
this.value before update: 100
this.value after update: 1000
this.value before proc: 100
this.value after proc: 1000
```

File: ios/Nodes/REAParamNode.m (hunk 2 above)
Comment (chrfalch) [reply ios/Nodes/REAParamNode.m]: Just testing to see if the node exists..? :) Maybe not necessary?

Comment (chrfalch) [reply ios/Nodes/REAParamNode.m:44]: The problem still exist. The problem arises if we use the `set` node in different contexts (different call id) - then the memoized value will be set, but when someone tries to read it again the wrong memoized value will be read. 

I think that using the dictionary to store memoized values is the problem - it will eventually lead to situations where different values are stored for the same node - giving unpredictable results when reading nodes from different contexts.

**Comment by osdnk** [comment]
```
import React, { Component } from 'react';
import { View } from 'react-native';
import Animated from 'react-native-reanimated';

const { set, call, block, proc } = Animated;

const updateValue = proc((a) => block([
  set(a, 1000),
]));

const updateValue2 = proc((a) => block([
  updateValue(a)
]));

const updateValue3 = proc((a) => block([
  updateValue2(a)
]));

export default class Example extends Component {
  constructor(props) {
    super(props);
    this.value1 = new Animated.Value(100);
    this.value2 = new Animated.Value(100);
  }

  render() {
    return (
      <View>
        <Animated.Code>
          {() => block([
            // First version without proc
            call(
              [this.value1],
              ([value]) => {
                console.log("this.value before update:", value);
              }
            ),
            set(this.value1, 1000),
            call(
              [this.value1],
              ([value]) => {
                console.log("this.value after update:", value);
              }
            ),
            // With proc
            call(
              [this.value2],
              ([value]) => {
                console.log("this.value before proc:", value);
              }
            ),
            updateValue3(this.value2),
            call(
              [this.value2],
              ([value]) => {
                console.log("this.value after proc:", value);
              }
            )])
          }
        </Animated.Code>
      </View>
    );
  }
}

```
After my fixes this works as expected
